Support record_shapes in RPC profiling

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44419 Support record_shapes in RPC profiling**

Closes https://github.com/pytorch/pytorch/issues/39969

This PR adds support for propagation of input shapes over the wire when the profiler is invoked with `record_shapes=True` over RPC. Previously, we did not respect this argument.

This is done by saving the shapes as an ivalue list and recovering it as the type expected (`std::vector<std::vector<int>>` on the client). Test is added to ensure that remote ops have the same `input_shapes` as if the op were run locally.

Remote ops profiler output before (run with `group_by_input_shapes=True`):

```
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::add        0.18%            67.229us         0.21%            76.449us         76.449us         1                2                []
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::empty      0.03%            9.220us          0.03%            9.220us          9.220us          1                2                []
```

After:
```
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::add        0.14%            91.001us         0.16%            102.094us        102.094us        1                2                [[100], [100], []]
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::empty      0.02%            11.093us         0.02%            11.093us         11.093us         1                2                [[], [], [], [], [], []]
```

Differential Revision: [D23591274](https://our.internmc.facebook.com/intern/diff/D23591274/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23591274/)!

rohan-varma committed:
Support record_shapes in RPC profiling 

Closes https://github.com/pytorch/pytorch/issues/39969

This PR adds support for propagation of input shapes over the wire when the profiler is invoked with `record_shapes=True` over RPC. Previously, we did not respect this argument.

This is done by saving the shapes as an ivalue list and recovering it as the type expected (`std::vector<std::vector<int>>` on the client). Test is added to ensure that remote ops have the same `input_shapes` as if the op were run locally.

Differential Revision: [D23591274](https://our.internmc.facebook.com/intern/diff/D23591274/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23591274/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support record_shapes in RPC profiling "


Closes https://github.com/pytorch/pytorch/issues/39969

This PR adds support for propagation of input shapes over the wire when the profiler is invoked with `record_shapes=True` over RPC. Previously, we did not respect this argument.

This is done by saving the shapes as an ivalue list and recovering it as the type expected (`std::vector<std::vector<int>>` on the client). Test is added to ensure that remote ops have the same `input_shapes` as if the op were run locally.

Differential Revision: [D23591274](https://our.internmc.facebook.com/intern/diff/D23591274/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23591274/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support record_shapes in RPC profiling "


Closes https://github.com/pytorch/pytorch/issues/39969

This PR adds support for propagation of input shapes over the wire when the profiler is invoked with `record_shapes=True` over RPC. Previously, we did not respect this argument.

This is done by saving the shapes as an ivalue list and recovering it as the type expected (`std::vector<std::vector<int>>` on the client). Test is added to ensure that remote ops have the same `input_shapes` as if the op were run locally.

Differential Revision: [D23591274](https://our.internmc.facebook.com/intern/diff/D23591274/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23591274/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support record_shapes in RPC profiling "


Closes https://github.com/pytorch/pytorch/issues/39969

This PR adds support for propagation of input shapes over the wire when the profiler is invoked with `record_shapes=True` over RPC. Previously, we did not respect this argument.

This is done by saving the shapes as an ivalue list and recovering it as the type expected (`std::vector<std::vector<int>>` on the client). Test is added to ensure that remote ops have the same `input_shapes` as if the op were run locally.

Differential Revision: [D23591274](https://our.internmc.facebook.com/intern/diff/D23591274/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23591274/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support record_shapes in RPC profiling "


Closes https://github.com/pytorch/pytorch/issues/39969

This PR adds support for propagation of input shapes over the wire when the profiler is invoked with `record_shapes=True` over RPC. Previously, we did not respect this argument.

This is done by saving the shapes as an ivalue list and recovering it as the type expected (`std::vector<std::vector<int>>` on the client). Test is added to ensure that remote ops have the same `input_shapes` as if the op were run locally.

Differential Revision: [D23591274](https://our.internmc.facebook.com/intern/diff/D23591274/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23591274/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Support record_shapes in RPC profiling "


Closes https://github.com/pytorch/pytorch/issues/39969

This PR adds support for propagation of input shapes over the wire when the profiler is invoked with `record_shapes=True` over RPC. Previously, we did not respect this argument.

This is done by saving the shapes as an ivalue list and recovering it as the type expected (`std::vector<std::vector<int>>` on the client). Test is added to ensure that remote ops have the same `input_shapes` as if the op were run locally.

Differential Revision: [D23591274](https://our.internmc.facebook.com/intern/diff/D23591274/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23591274/)!

[ghstack-poisoned]

pritamdamania87 commented:
Could you paste the before and after autograd profiler results to the PR summary so its easy to see the appropriate changes in the profiler output?

pritamdamania87 reviewed (commented):
Unit test looks good to me, will let @ilia-cher sign off on the profiler.cpp changes.

rohan-varma commented:
> Could you paste the before and after autograd profiler results to the PR summary so its easy to see the appropriate changes in the profiler output?

@pritamdamania87 The before/after of the RPC remote events are in the summary now.

rohan-varma committed:
Update on "Support record_shapes in RPC profiling "


Closes https://github.com/pytorch/pytorch/issues/39969

This PR adds support for propagation of input shapes over the wire when the profiler is invoked with `record_shapes=True` over RPC. Previously, we did not respect this argument.

This is done by saving the shapes as an ivalue list and recovering it as the type expected (`std::vector<std::vector<int>>` on the client). Test is added to ensure that remote ops have the same `input_shapes` as if the op were run locally.

Remote ops profiler output before (run with `group_by_input_shapes=True`):

```
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::add        0.18%            67.229us         0.21%            76.449us         76.449us         1                2                []
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::empty      0.03%            9.220us          0.03%            9.220us          9.220us          1                2                []
```

After:
```
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::add        0.14%            91.001us         0.16%            102.094us        102.094us        1                2                [[100], [100], []]
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::empty      0.02%            11.093us         0.02%            11.093us         11.093us         1                2                [[], [], [], [], [], []]
```

Differential Revision: [D23591274](https://our.internmc.facebook.com/intern/diff/D23591274/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23591274/)!

[ghstack-poisoned]

ilia-cher commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -516,14 +517,34 @@ void Event::record(bool record_cuda) {
+  shapesFromServer.reserve(shapeList.size());
+  for (size_t i = 0 ; i < shapeList.size(); ++i) {
+    std::vector<int64_t> s;
+    auto curShapesList = shapeList.get(i).toList();
```
should we also check `isList` here?

ilia-cher commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -516,14 +517,34 @@ void Event::record(bool record_cuda) {
       " elements to reconstruct Event.");
 
+  // Reconstruct input shapes from ivalues.
+  auto shapeListFromServer = ivalues.get(EventIValueIdx::SHAPES);
```
nit: could you remove mentions of server?

rohan-varma replied:
Removed all of the mentions.

rohan-varma committed:
Update on "Support record_shapes in RPC profiling "


Closes https://github.com/pytorch/pytorch/issues/39969

This PR adds support for propagation of input shapes over the wire when the profiler is invoked with `record_shapes=True` over RPC. Previously, we did not respect this argument.

This is done by saving the shapes as an ivalue list and recovering it as the type expected (`std::vector<std::vector<int>>` on the client). Test is added to ensure that remote ops have the same `input_shapes` as if the op were run locally.

Remote ops profiler output before (run with `group_by_input_shapes=True`):

```
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::add        0.18%            67.229us         0.21%            76.449us         76.449us         1                2                []
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::empty      0.03%            9.220us          0.03%            9.220us          9.220us          1                2                []
```

After:
```
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::add        0.14%            91.001us         0.16%            102.094us        102.094us        1                2                [[100], [100], []]
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::empty      0.02%            11.093us         0.02%            11.093us         11.093us         1                2                [[], [], [], [], [], []]
```

Differential Revision: [D23591274](https://our.internmc.facebook.com/intern/diff/D23591274/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23591274/)!

[ghstack-poisoned]

ilia-cher reviewed (approved):
LG!

rohan-varma committed:
Update on "Support record_shapes in RPC profiling "


Closes https://github.com/pytorch/pytorch/issues/39969

This PR adds support for propagation of input shapes over the wire when the profiler is invoked with `record_shapes=True` over RPC. Previously, we did not respect this argument.

This is done by saving the shapes as an ivalue list and recovering it as the type expected (`std::vector<std::vector<int>>` on the client). Test is added to ensure that remote ops have the same `input_shapes` as if the op were run locally.

Remote ops profiler output before (run with `group_by_input_shapes=True`):

```
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::add        0.18%            67.229us         0.21%            76.449us         76.449us         1                2                []
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::empty      0.03%            9.220us          0.03%            9.220us          9.220us          1                2                []
```

After:
```
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::add        0.14%            91.001us         0.16%            102.094us        102.094us        1                2                [[100], [100], []]
rpc_sync#aten::add(worker1 -> worker2)#remote_op: aten::empty      0.02%            11.093us         0.02%            11.093us         11.093us         1                2                [[], [], [], [], [], []]
```

Differential Revision: [D23591274](https://our.internmc.facebook.com/intern/diff/D23591274/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23591274/)!

[ghstack-poisoned]
